### ISSUE TITLE
Runtime MatchError when using countDistinct

### ISSUE THREAD

**Issue by Frumsz** [slick/slick#1117]
I've come accross what I presume to be a bug, or otherwise maybe an improvement (since I could see it making sense having to specify a column(s) to distinct on):

I have encountered this with RC-1 and RC-2 and Pgsql

Consider the following:
```scala
val products = TableQuery[Products]
val customersToProducts = TableQuery[CustomersToProducts]

val productsWithCustomerCountQuery = products.map(p => (p, customersToProducts.filter(_.productId === p.id).countDistinct))

db.run(productsWithCustomerCountQuery.result)
```
The query will result in a runtime match error:
```scala
play.api.Application$$anon$1: Execution exception[[MatchError: Some(Pure $@1677218149) (of class scala.Some)]]
        at play.api.Application$class.handleError(Application.scala:296) ~[play_2.11-2.3.8.jar:2.3.8]
        at play.api.DefaultApplication.handleError(Application.scala:402) [play_2.11-2.3.8.jar:2.3.8]
        at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$14$$anonfun$apply$1.applyOrElse(PlayDefaultUpstreamHandler.scala:205) [play_2.11-2.3.8.jar:2.3.8]
        at play.core.server.netty.PlayDefaultUpstreamHandler$$anonfun$14$$anonfun$apply$1.applyOrElse(PlayDefaultUpstreamHandler.scala:202) [play_2.11-2.3.8.jar:2.3.8]
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36) [scala-library-2.11.6.jar:na]
Caused by: scala.MatchError: Some(Pure $@1677218149) (of class scala.Some)
        at slick.compiler.FuseComprehensions.slick$compiler$FuseComprehensions$$tr$1(Relational.scala:361) ~[slick_2.11-3.0.0-RC2.jar:na]
        at slick.compiler.FuseComprehensions$$anonfun$slick$compiler$FuseComprehensions$$tr$1$2.apply(Relational.scala:372) ~[slick_2.11-3.0.0-RC2.jar:na]
        at slick.compiler.FuseComprehensions$$anonfun$slick$compiler$FuseComprehensions$$tr$1$2.apply(Relational.scala:372) ~[slick_2.11-3.0.0-RC2.jar:na]
        at slick.ast.Util$$anonfun$mapOrNone$1.apply(Util.scala:15) ~[slick_2.11-3.0.0-RC2.jar:na]
        at slick.ast.Util$$anonfun$mapOrNone$1.apply(Util.scala:14) ~[slick_2.11-3.0.0-RC2.jar:na]
```

**Comment by stempler** [comment]
Also just encountered this runtime error with Slick 2.0.1 and the PostgreSQL driver.
As @Frumsz suggested works if mapping the query to a specific column before doing countDistinct.

**Comment by anand-singh** [comment]
+1

**Comment by BartoszJanota** [comment]
Hi guys,

I encountered the same problem and I solved this the same way as @stempler suggested.

@Frumsz consider something like:

```Scala
val productsWithCustomerCountQuery = products.map(p => (p, customersToProducts.filter(_.productId === p.id).map(x => x.productId).countDistinct))

db.run(productsWithCustomerCountQuery.result)
```

**Comment by szeiger** [comment]
Fixed in 3.1
